lib/core-client/src/preview/normalizeStory.ts

```diff
@@ -0,0 +1,110 @@
+import { logger } from '@storybook/client-logger';
+import { ClientApi } from '@storybook/client-api';
+import { storyNameFromExport, toId } from '@storybook/csf';
+import dedent from 'ts-dedent';
+import deprecate from 'util-deprecate';
+import global from 'global';
+
+const { PREVIEW_CSF_V3 } = global;
+const isCsfV3 = () => {
+  try {
+    return !!PREVIEW_CSF_V3;
+  } catch (e) {
+    return false;
+  }
+};
+
+const deprecatedStoryAnnotation = dedent`
+CSF .story annotations deprecated; annotate story functions directly:
+- StoryFn.story.name => StoryFn.storyName
+- StoryFn.story.(parameters|decorators) => StoryFn.(parameters|decorators)
+See https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#hoisted-csf-annotations for details and codemod.
+`;
+
+const deprecatedStoryAnnotationWarning = deprecate(() => {}, deprecatedStoryAnnotation);
+
+/**
+ * Utilities for normalizing a story to support different
+ */
+export const normalizeV2 = (key: string, storyExport: any, meta: any, globalRender: any) => {
+  const exportType = typeof storyExport;
+  if (exportType !== 'function') {
+    logger.info(`Unexpected story export "${key}": expected function, received "${exportType}."`);
+  }
+
+  const storyFn = storyExport;
+  const { story } = storyFn;
+  const { storyName = story?.name } = storyFn;
+
+  // storyFn.x and storyFn.story.x get merged with
+  // storyFn.x taking precedence in the merge
+  const storyParams = { ...story?.parameters, ...storyFn.parameters };
+  const decorators = [...(storyFn.decorators || []), ...(story?.decorators || [])];
+  const loaders = [...(storyFn.loaders || []), ...(story?.loaders || [])];
+  const args = { ...story?.args, ...storyFn.args };
+  const argTypes = { ...story?.argTypes, ...storyFn.argTypes };
+
+  if (story) {
+    logger.debug('deprecated story', story);
+    deprecatedStoryAnnotationWarning();
+  }
+
+  const exportName = storyNameFromExport(key);
+  const parameters = {
+    ...storyParams,
+    __id: toId(meta.id || meta.title, exportName),
+    decorators,
+    loaders,
+    args,
+    argTypes,
+  };
+
+  return {
+    name: storyName || exportName,
+    storyFn,
+    parameters,
+  };
+};
+
+export const normalizeV3 = (key: string, storyExport: any, meta: any, globalRender: any) => {
```

Comment:
Is it possible to type them better? Something like:
```suggestion
export const normalizeV3 = (key: string, storyExport: any, meta: Meta, globalRender: Story) => {
```